solr/core/src/java/org/apache/solr/cluster/events/impl/ClusterEventProducerFactory.java

```diff
@@ -0,0 +1,173 @@
+      clusterEventProducer.setDelegate((ClusterEventProducer) clusterEventProducerInfo.getInstance());
+    } else {
+      // use the default NoOp impl
+    }
+    // transfer those listeners that were already registered to the initial impl
+    transferListeners(clusterEventProducer, plugins);
+    // install plugin registry listener
+    ContainerPluginsRegistry.PluginRegistryListener pluginListener = new ContainerPluginsRegistry.PluginRegistryListener() {
+      @Override
+      public void added(ContainerPluginsRegistry.ApiInfo plugin) {
+        if (plugin == null || plugin.getInstance() == null) {
+          return;
+        }
+        Object instance = plugin.getInstance();
+        if (instance instanceof ClusterEventListener) {
+          ClusterEventListener listener = (ClusterEventListener) instance;
+          clusterEventProducer.registerListener(listener);
+        } else if (instance instanceof ClusterEventProducer) {
+          // replace the existing impl
+          if (cc.getClusterEventProducer() instanceof DelegatingClusterEventProducer) {
+            ((DelegatingClusterEventProducer) cc.getClusterEventProducer())
+                .setDelegate((ClusterEventProducer) instance);
+          } else {
+            log.warn("Can't configure plugin-based ClusterEventProducer while CoreContainer is still loading - " +
+                " using existing implementation {}", cc.getClusterEventProducer().getClass().getName());
+          }
+        }
+      }
+
+      @Override
+      public void deleted(ContainerPluginsRegistry.ApiInfo plugin) {
+        if (plugin == null || plugin.getInstance() == null) {
+          return;
+        }
+        Object instance = plugin.getInstance();
+        if (instance instanceof ClusterEventListener) {
+          ClusterEventListener listener = (ClusterEventListener) instance;
+          clusterEventProducer.unregisterListener(listener);
+        } else if (instance instanceof ClusterEventProducer) {
+          // replace the existing impl with NoOp
+          if (cc.getClusterEventProducer() instanceof DelegatingClusterEventProducer) {
+            ((DelegatingClusterEventProducer) cc.getClusterEventProducer())
+                .setDelegate(new NoOpProducer(cc));
+          } else {
+            log.warn("Can't configure plugin-based ClusterEventProducer while CoreContainer is still loading - " +
+                " using existing implementation {}", cc.getClusterEventProducer().getClass().getName());
+          }
+        }
+      }
+
+      @Override
+      public void modified(ContainerPluginsRegistry.ApiInfo old, ContainerPluginsRegistry.ApiInfo replacement) {
+        added(replacement);
+        deleted(old);
+      }
+    };
+    plugins.registerListener(pluginListener);
+    created = true;
+    return clusterEventProducer;
+  }
+
+  private void transferListeners(ClusterEventProducer target, ContainerPluginsRegistry plugins) {
```

Comment:
Would there be a way to register listeners into an intermediate object (not tied to any specific producer), then have event producers feed events into this object (conceptually it would behave like a queue, but the implementation can be direct listener method calls). This would save having methods such as this one (and would not require having a producer in order to register listeners, if that helps...).